core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchFragment.kt

```diff
@@ -104,13 +113,71 @@ class SearchFragment : BaseFragment() {
       adapter = searchAdapter
       layoutManager = LinearLayoutManager(context, RecyclerView.VERTICAL, false)
       setHasFixedSize(true)
+      // Add scroll listener to detect when the last item is reached
+      addOnScrollListener(object : RecyclerView.OnScrollListener() {
+        override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
+          super.onScrolled(recyclerView, dx, dy)
+
+          val layoutManager = recyclerView.layoutManager as LinearLayoutManager
+          val totalItemCount = layoutManager.itemCount
+          val lastVisibleItem = layoutManager.findLastVisibleItemPosition()
+          // Check if the user is about to reach the last item
+          if (!isDataLoading && totalItemCount <= lastVisibleItem + VISIBLE_ITEMS_THRESHOLD) {
+            // Load more data when the last item is almost visible
+            loadMoreSearchResult()
+          }
+        }
+      })
     }
     lifecycleScope.launchWhenCreated {
       searchViewModel.effects.collect { it.invokeWith(this@SearchFragment.coreMainActivity) }
     }
     handleBackPress()
   }
 
+  /**
+   * Loads more search results if available.
+   * If data is currently being loaded, the method returns early and does not initiate another load request.
+   * The loading process sets the `isDataLoading` flag to true to prevent multiple load requests at once.
+   * The method fetches more search results based on the current `searchState` and the `safeStartIndex`,
+   * which represents the current item count in the `searchAdapter`.
+   * If there are no more results to fetch, the method sets `isDataLoading` to true to avoid unnecessary data loading.
+   * If there are new results available, it checks for duplicate entries, especially for search history items.
+   * Non-duplicate results are appended to the existing dataset in the `searchAdapter`.
+   * If new data (non-duplicate items) is loaded successfully, the method sets `isDataLoading` to false.
+   * If there are no new non-duplicate results, the method sets `isDataLoading` to true
+   * to avoid unnecessary data loading, as there are no more items available.
+   */
+  private fun loadMoreSearchResult() {
+    if (isDataLoading) return
+    val safeStartIndex = searchAdapter?.itemCount ?: 0
+    isDataLoading = true
```

Comment:
We should extract this data logic to repostiory classes to make separate concerns.

Reply:
Sorry, I didn't get you, are you talking about `val safeStartIndex = searchAdapter?.itemCount ?: 0` this?

Reply:
No, the logic of the function shouldn't be in the `SearchFragment`. We should refactor to the data classes maybe viewmodel or repository depending on the situation for separate concerns.

Reply:
@gouri-panda Done.